### PR TITLE
Add the ability to migrate CoreDNS configmap in kube-up

### LINKED ISSUE #78123: [Failing tests] Multiple job failures on the BeforeSuite step while waiting for CoreDNS to be ready
**Which jobs are failing**:
Multiple jobs on `master-informing`.

**Which test(s) are failing**:
`Before Suite` on the following jobs:
- https://testgrid.k8s.io/sig-release-master-informing#kubeadm-kinder-upgrade-stable-master
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-master
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-master-parallel
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-cluster-parallel
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-cluster
- https://testgrid.k8s.io/sig-release-master-informing#gce-master-new-downgrade-cluster
- https://testgrid.k8s.io/sig-release-master-informing#gce-master-new-downgrade-cluster-parallel
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-cluster-new
- https://testgrid.k8s.io/sig-release-master-informing#gce-new-master-upgrade-cluster-new-parallel

**Since when has it been failing**:
Since 5/17.

**Testgrid links**:
See above.

**Reason for failure**:
All the failures mentioned above seem to have to have the same root cause, namely failure while waiting for CoreDNS to be ready.
Example: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-upgrade-stable-master/1129638974428549121
```
May 18 07:09:00.075: Error waiting for all pods to be running and ready: 2 / 14 pods in namespace "kube-system" are NOT in RUNNING and READY state in 10m0s
POD                      NODE                   PHASE   GRACE CONDITIONS
coredns-65546fffc9-j9p6b kinder-upgrade-worker2 Running       [{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2019-05-18 06:52:59 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2019-05-18 06:52:59 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [coredns]} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2019-05-18 06:52:59 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [coredns]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2019-05-18 06:52:59 +0000 UTC Reason: Message:}]
```

/sig testing
/priority critical-urgent
/kind failing test
/milestone v1.15
/cc @jimangel @alejandrox1 @rarchk @alenkacz

### PR DESCRIPTION
**Pull request by rajansandeep**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Following CoreDNS ConfigMap changes in #78030, it is required that the ConfigMap gets updated during an `upgrade`. 
If I understand correctly, `EnsureExists` doesn't update the ConfigMap during an upgrade and requires `Reconcile` for the ConfigMap to upgrade.

This is evident from the logs:

Logs from `ci-kubernetes-e2e-gce-new-master-upgrade-cluster`:
```
event for coredns-5cd44cdc5d-lzbmv: {kubelet bootstrap-e2e-minion-group-b8n6} Unhealthy: Readiness probe failed: Get http://10.64.2.20:8181/ready: dial tcp 10.64.2.20:8181: connect: connection refused
```
Logs from `ci-kubernetes-e2e-gce-master-new-downgrade-cluster`:
```
/etc/coredns/Corefile:4 - Error during parsing: Unknown directive 'ready'
```
In both cases, the CoreDNS deployment is updated but not the ConfigMap.
This PR aims to add the ability to migrate the CoreDNS ConfigMap in kube-up in the event of an upgrade/downgrade.

It also bumps the version of CoreDNS to 1.5.0

**Which issue(s) this PR fixes**:

Fixes #78123 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by MrHohn** [comment]
@rajansandeep IIUC if we set this to reconcile, user will no longer be able to [customize CoreDNS service](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#coredns-configuration-equivalent-to-kube-dn) because whatever change they make will be reconciled?

**Comment by chrisohaver** [comment]
@mrhohn, a user would have to customize their configmap via the addon manager yaml definitions, not directly in the configmap object, just as they would do to edit the Deployment for example.

If that’s not reasonable, perhaps instead we could simply add a step to delete the configmap near the end of upgrade.sh, so it gets recreated by addon manager using the new manifest.

However, If the upgrade.sh script is used in production gce clusters, then maintaining old customizations is important, and we should probably integrate the coredns config migration tools into the upgrade.sh script, as @rajansandeep did in kubeadm.

**Comment by chrisohaver** [comment]
Thinking on this further...

If we use `Reconcile`, then some DNS tests may become flakey (any ones that change the configmap) - since the configmap could get reconciled before the tests complete.

I think the safer route would be to leave it as `EnsureExists` and either delete the configmap so the addon-manager recreates it, or directly update the Configmap with the new corefile.  Unless there is a more direct way of telling the addon-manager to reconcile a specific resource on demand.

This should be followed by a loop/check to ensure that CoreDNS has received and loaded the new corefile.

**Comment by MrHohn** [comment]
> @MrHohn, a user would have to customize their configmap via the addon manager yaml definitions, not directly in the configmap object, just as they would do to edit the Deployment for example.

Unfortunately that is not possible in some cloud environment (e.g. GKE) where master is managed by provider. I remembered that is one of the reasons why we surfaced these configurations from out of the original deployment file?

> However, If the upgrade.sh script is used in production gce clusters, then maintaining old customizations is important

AFAIK this upgrade.sh script isn't used in production, so adding special steps there as workaround SGTM.

And +1 to the further thoughts.. There may be side effects for changing to `Reconcile`.

**Comment by MrHohn** [comment]
Though thinking it is still important to understand how a configuration change like this should happen when upgrading cluster. As kubeadm has special steps to handle this, should we expect each cloudprovider to support similar stuff in their upgrade pipeline or should we investigate a more generic approach?

**Comment by chrisohaver** [comment]
> Should we expect each cloudprovider to support similar stuff in their upgrade pipeline

That was my thinking for now - hence the [coredns migration library](https://github.com/coredns/deployment/tree/master/kubernetes/migration).  My thinking is that various cluster mgmt tools could use this to migrate coredns configs (following the example in kubeadm).

> should we investigate a more generic approach

We can, possibly for 1.16 ... There may have been config migrations required for other components (etcd?), if so would be good to see how those are handled.

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [inline cluster/gce/upgrade.sh]: I think this should say "Waiting for Addon Manager to restore CoreDNS ConfigMap"

**Review by chrisohaver — COMMENTED** [review]
Looks good.  One minor wording comment...

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: Just sanity check --- do we have any upgrade suite that runs with kube-dns, in which case addon manager might never restore this?

**Review by MrHohn — APPROVED** [review]
/lgtm

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: Actually, mind putting a comment explaining why this step is needed?

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [reply cluster/gce/upgrade.sh]: good point.  The kube-dns scenario slipped my mind.  We should only call `update-coredns-config` if coredns is set as the cluster dns.  This is stored in an env variable during upgrade (i recall/hope).

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: Done!

**Review by chrisohaver — APPROVED** [review]
LGTM! Thanks!

**Review by MrHohn — APPROVED** [review]
/lgtm
/approve

**Comment by rajansandeep** [comment]
@dstrebel @smourapina @alejandrox1 for milestone 1.15 tag
/cc @bowei for final approval

**Comment by rajansandeep** [comment]
/cc  @kubernetes/release-team 
Needs an approval tag to go through

**Comment by MrHohn** [comment]
This as a bug fix shouldn't be impacted by code-freeze right?

(Unfortunately I don't have approval on this.)
/approve

**Comment by rajansandeep** [comment]
@MrHohn I’m not sure, but this is a topic of discussion at sig-release. 
@bowei can approve this but I’m not able to reach him.

**Comment by MrHohn** [comment]
@rajansandeep Bowei might not be available this week. Checking if @yujuhong can help approve this if it makes sense :)

File: cluster/gce/upgrade.sh
Comment (liggitt) [inline cluster/gce/upgrade.sh]: isn't this data loss? I thought that configmap contained user data

**Comment by liggitt** [comment]
/hold

for question on whether this is deleting user data

File: cluster/gce/upgrade.sh
Comment (MrHohn) [reply cluster/gce/upgrade.sh]: Yep this may contains user data. We were under the impression that this script is not used in production in any way. Otherwise we will need something like https://github.com/kubernetes/kubernetes/pull/78033 in the script. Or maybe do what was recommended on https://github.com/kubernetes/kubernetes/pull/78033#issuecomment-498328908.

Comment (liggitt) [reply cluster/gce/upgrade.sh]: oh. I have no idea what calls this script, will defer to cluster/gce owners

Comment (liggitt) [reply cluster/gce/upgrade.sh]: > We were under the impression that this script is not used in production in any way

if that turns out to be true, and we do go forward with this change, there needs to be a comment here documenting that this is known to delete a configmap that contains user data, and that it is ok because this script is not used by any production setups. I could easily see this being copied-pasted as "the way" to clear/upgrade the dns configmap

**Comment by chrisohaver** [comment]
User data deleted or not, with the revert of CoreDNS back to 1.3.1, this is no longer needed.

**Comment by xmudrii** [comment]
@chrisohaver @liggitt Hello! I'm 1.15 Bug triage shadow and I'd like to verify the status of this PR. The code freeze has started last week, which means only release-blocking PRs are considered to be merged.

Do we still want to merge it for 1.15 or postponing it to 1.16 is okay?

**Comment by neolit123** [comment]
@MrHohn @chrisohaver @rajansandeep 
the kubeadm revert to coredns 1.3.1 merged:
https://github.com/kubernetes/kubernetes/pull/78545

it was not clear to me if /cluster will be happy with the 1.5.x bump, thus https://github.com/kubernetes/kubernetes/pull/78545 did not touch /cluster.

**Comment by MrHohn** [comment]
My preference (not very strong) is reverting the /cluster part as well to avoid causing potential breakage when user upgrade and to be consistent.

Will let @rajansandeep @chrisohaver decide.

**Comment by chrisohaver** [comment]
IMO, it makes sense to test using the version that most people will be using with k8s 1.15, which would be coredns 1.3.1. So that means reverting the cluster addon manifests, and associated fixes that have already merged.



On Jun 3, 2019, at 8:39 PM, Zihong Zheng <notifications@github.com<mailto:notifications@github.com>> wrote:


My preference (not very strong) is reverting the /cluster part as well to avoid causing potential breakage when user upgrade and to be consistent.

Will let @rajansandeep<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_rajansandeep&d=DwMCaQ&c=CWsWqoUynJrLESJQduOsxQ&r=fqtOgeJR6kXT8siDsioK_cuszjJxxWjoOLsT1wkHy5k&m=QBLXPX_XKMhHpLkw0RFDXL4W6PN0Xumw8Et_oHcRULM&s=NxxXcQeonaZ6nkNJcEySHQwIwYlxzzjn-2xfGLA59oI&e=> @chrisohaver<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_chrisohaver&d=DwMCaQ&c=CWsWqoUynJrLESJQduOsxQ&r=fqtOgeJR6kXT8siDsioK_cuszjJxxWjoOLsT1wkHy5k&m=QBLXPX_XKMhHpLkw0RFDXL4W6PN0Xumw8Et_oHcRULM&s=xAPskCXAfSLgJynTH7SXluBy8jQ7EOHCvh9g1XRhJUs&e=> decide.

**Comment by alejandrox1** [comment]
Hi @rajansandeep , will this PR still solve the failures on the upgrade tests with the revert of coreDNS to 1.3.1? It looks to me like cluster/gce/* is not affected by changes in kubeadm but im not very familiar with this part of the repo.

**Comment by rajansandeep** [comment]
This PR is required for CoreDNS 1.5.0. The upgrade failures will no longer occur once CoreDNS is reverted back to 1.3.1 (https://github.com/kubernetes/kubernetes/pull/78691)

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: @liggitt I have added comments to warn that this shouldn't be followed in a production setup. Let me know if this is acceptable.

We still need to verify whether this script is used in production or not.

**Comment by chrisohaver** [comment]
Update: there is resistance to reverting #78691, because upgrading CoreDNS 1.5.0 may have inadvertently resolved a scale issue observed with CoreDNS 1.3.1 and the new version of COS (root cause not verified/determined yet).

So, aiming again to get this fixed merged.

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [reply cluster/gce/upgrade.sh]: @bowei, is `cluster/gce/upgrade.sh` prescribed for use in production clusters in any way?

**Comment by liggitt** [comment]
Adding to the milestone for visibility until the discussion is resolved

/milestone v1.15

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [reply cluster/gce/upgrade.sh]: Searching the kubernetes repo, I only see use of this script from within the e2e/tests.

However, I found references in kubernetes/website ... in this page ...

https://kubernetes.io/docs/tasks/administer-cluster/cluster-management/

> Upgrades on open source Google Compute Engine (GCE) clusters are controlled by the cluster/gce/upgrade.sh script.
> Get its usage by running cluster/gce/upgrade.sh -h.

So, apparently we _do_ prescribe its usage in production.

Which means we cannot merge this as is, which will tumble a few more dominoes...

**Comment by soggiest** [comment]
Any updates on this issue? @timothysc any thoughts?

**Comment by spiffxp** [comment]
my personal preference is that we not try a half-measure for 1.15, and instead take our time to develop a proper upgrade path for coredns as part of 1.16 (ref: https://github.com/kubernetes/kubernetes/pull/78691#issuecomment-500003085)

**Comment by alejandrox1** [comment]
moving this from the milestone so that we can work on a proper upgrade path with coredns v1.15, xref: https://github.com/kubernetes/kubernetes/pull/78691#issuecomment-500003085

/milestone v1.16

**Comment by rajansandeep** [comment]
Since it’s unclear whether this is used in production or not, I’ll update this PR soon to make sure that the CoreDNS configmap is migrated instead of just deleting the configmap and being reconciled by the add on manager. It will follow the same migration logic as for kubeadm in #78033.

**Comment by rajansandeep** [comment]
@MrHohn @liggitt I've updated the PR to gracefully migrate the CoreDNS ConfigMap for the upgrade. PTAL and let me know if this is a step in the right direction.

**Comment by rajansandeep** [comment]
@kubernetes/sig-cluster-lifecycle-pr-reviews 
Need eyes on this so that it doesn't end up getting too late to get this in 1.16!

**Comment by neolit123** [comment]
ping @MrHohn 
PTAL at the latest update on the `/cluster/*` change for the CoreDNS bump.

File: cluster/addons/dns/coredns/coredns.yaml.base
Comment (MrHohn) [inline cluster/addons/dns/coredns/coredns.yaml.base]: Probably better to split the CoreDNS bump into a separate commit?

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: QQ: would the script stops if the `mktemp` command fails?

(Just hoping `rm -rf ${download_dir}` below will not mess up something else :) )

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: Humm.. It seems like we only wait for kube-apiserver to respond before calling `update-coredns-config()`. It may take a bit longer for CoreDNS deployment to be updated in practice. Anyway that we can know for sure the deployment is the updated one?

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: This line is hard to notice --- can we move this down to after all the funcs definition? And logging around this might be helpful when debugging.

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: By simply applying the downloaded config, does it mean we don't preserve user setup on downgrade (or version unchanged)?

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: Probably doesn't matter --- why `replace` v.s. `update`?

File: cluster/gce/upgrade.sh
Comment (MrHohn) [inline cluster/gce/upgrade.sh]: `gcloud scp` may be better for coping file?

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [reply cluster/gce/upgrade.sh]: Yes.  A downgrade re-deploys the default configuration for the earlier version.
This always results in a functioning cluster DNS.  If instead, the custom Corefile was retained, there is a chance that the Corefile uses an option or plugin that the older version of CoreDNS will not recognize (causing CoreDNS to exit).

Downward migrations are not fully supported by the Corefile migration package.  Although it is technically possible to support then, they are complex, and supporting them would create more maintenance burden.  So, I'm hoping that supporting downward migrations is not required.

IMO, re-deploying the default configuration for a downgrade is acceptable. Downgrades are much less common.  If a user needs to downgrade a system with a custom Corefile, they will need to manually modify the resulting (default) Corefile after the downgrade is complete. The hands-on nature can be documented in the k8.io doc page.

In the past, retaining custom configuration during downgrades has not been a concern in the context of this script. For example when downgrading from a version of k8s that used CoreDNS to a version that used kube-dns, custom configurations were not maintained (kube-dns defaults deployed).

Comment (MrHohn) [reply cluster/gce/upgrade.sh]: @chrisohaver Thanks for your insights, that explains the reason behind clearly. Yeah re-deploy the default config on downgrade and add doc/comment around this behavior sounds totally fine to me.

File: cluster/addons/dns/coredns/coredns.yaml.base
Comment (rajansandeep) [reply cluster/addons/dns/coredns/coredns.yaml.base]: Done

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: I am not sure of a way to know this. It's the reason I've added `sleep` before fetching the version.

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: I don't this `kubectl update` exists. Hence I had to use `replace`

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: yes! done.

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: I added a comment in the script. We can also add a doc wherever applicable.

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: done

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: I added an `exit` in case the command fails

File: cluster/gce/upgrade.sh
Comment (chrisohaver) [reply cluster/gce/upgrade.sh]: > I added a comment in the script. We can also add a doc wherever applicable.

I guess it would go in ...
https://kubernetes.io/docs/tasks/administer-cluster/cluster-management/#upgrading-google-compute-engine-clusters

File: cluster/gce/upgrade.sh
Comment (MrHohn) [reply cluster/gce/upgrade.sh]: Ah sorry, I meant to say `kubectl apply`

File: cluster/gce/upgrade.sh
Comment (MrHohn) [reply cluster/gce/upgrade.sh]: We can possibly retrieve the CoreDNS version (or `resourceVersion`, etc.) before and after the upgrade, and make sure it was changed?

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: `kubectl replace` destroys the previous configmap and creates a new one, while `kubectl apply` overwrites the existing one. I've changed it to use `kubectl apply`.

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: Thank you, I've added the check now.

**Review by MrHohn — APPROVED** [review]
Thanks, this LGTM.
/lgtm

File: cluster/gce/upgrade.sh
Comment (rajansandeep) [reply cluster/gce/upgrade.sh]: Should I add a log inside the loop or before it?

**Comment by rajansandeep** [comment]
@mrhohn Thank you for the reviews! I'd ask for one more favor for getting any one of the `cluster/gce/OWNERS` to give approval if possible. I remember I had a tough time last time. 

Pinging @BenTheElder @cblecker for `build/OWNERS` approval.

File: cluster/gce/upgrade.sh
Comment (MrHohn) [reply cluster/gce/upgrade.sh]: Right, I guess before the loop would be good enough.

**Comment by MrHohn** [comment]
We might need approval from @bowei for cluster/gce/OWNERS. Let me try one more time.
/approve

**Comment by MrHohn** [comment]
BTW would be great to squash the fixup commits.
/lgtm

**Comment by rajansandeep** [comment]
> BTW would be great to squash the fixup commits.

Done.

**Review by BenTheElder — APPROVED** [review]
/approve
for build

**Comment by rajansandeep** [comment]
Awaiting final approval from @bowei

**Comment by chrisohaver** [comment]
Is there a way to re-link this PR after the fork is recreated? (tried re-creating fork and pushing a cloned branch, but that didn't work)
Or is a new PR required?

On Aug 28, 2019, at 12:40 PM, Josh Berkus <notifications@github.com<mailto:notifications@github.com>> wrote:


hey, @rajansandeep<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_rajansandeep&d=DwMCaQ&c=CWsWqoUynJrLESJQduOsxQ&r=fqtOgeJR6kXT8siDsioK_cuszjJxxWjoOLsT1wkHy5k&m=eJzqwqkYBYrPKBkRLIX97AjmrZYi5qXr7SBnJcVmTSI&s=EqRTILRpxPNl3RIk3kr54bIwUHrzubPVKHdPuFHvbTc&e=> we can't accept this PR now because you deleted, or hid, your fork. Aside from Prow, I'm pretty sure there's legal issues from accepting an orphan PR. You're going to have to recreate it. :-(

**Review by bowei — COMMENTED** [review]
Is this the recommended way to do things? It seems like a vector to get remote binaries to run on your machine.

**Comment by rajansandeep** [comment]
closing in favor of https://github.com/kubernetes/kubernetes/pull/82093

**Pull request closed without merging**
